spec/compiler/semantic/proc_spec.cr

```diff
@@ -1119,4 +1119,20 @@ describe "Semantic: proc" do
       Foo.foo
     )) { proc_of int32 }
   end
+
+  it "doesn't cause upcast bug (#8428)" do
+    assert_type(%(
+      def foo
+        if true
+          begin
+             ->(){""}
+          rescue
+            return ->(){}
+          end
+        end
+      end
+
+      foo.not_nil!
```

Comment:
Note that `not_nil!` is provided by the std-lib which is not available in the semantic specs by default.

An alternative is to add an `else raise ""` in the `if true` of `foo`. That should get rid of the not_nil.

Otherwise the assert_type expected type can be changed.

Note that currently the spec fails but for the unwanted reason.